app/controllers/resources_controller.rb

```diff
@@ -1,17 +1,40 @@
 # frozen_string_literal: true
 
 class ResourcesController < ApplicationController
+  before_action :load_resource
+
   def show
-    @resource = find_resource(params[:id])
-    authorize @resource
+    authorize(@resource)
+  end
+
+  def download
+    if @resource.is_a?(WorkVersion)
+      redirect_to_s3
+    else
+      render file: Rails.root.join('public', '404.html'), status: :not_found
+    end
   end
 
   private
 
+    def load_resource
+      @resource = if params[:resource_id]
+                    find_resource(params[:resource_id])
+                  else
+                    find_resource(params[:id])
+                  end
```

Comment:
This kind of type checking would be unnecessary if we split downloads into its own `DownloadsController`